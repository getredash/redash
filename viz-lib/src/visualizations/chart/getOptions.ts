import { merge } from "lodash";
import { visualizationsSettings } from "@/visualizations/visualizationsSettings";

const DEFAULT_OPTIONS = {
  globalSeriesType: "column",
  sortX: true,
  legend: { enabled: true, placement: "auto", traceorder: "normal" },
  xAxis: { type: "-", labels: { enabled: true } },
  yAxis: [{ type: "linear" }, { type: "linear", opposite: true }],
  alignYAxesAtZero: false,
  error_y: { type: "data", visible: true },
  series: { stacking: null, error_y: { type: "data", visible: true } },
  seriesOptions: {},
  valuesOptions: {},
  columnMapping: {},
  direction: { type: "counterclockwise" },
  sizemode: "diameter",
  coefficient: 1,
  piesort: true,
  color_scheme: "Redash",

  // showDataLabels: false, // depends on chart type
  numberFormat: "0,0[.]00000",
  percentFormat: "0[.]00%",
  // dateTimeFormat: 'DD/MM/YYYY HH:mm', // will be set from visualizationsSettings
  textFormat: "", // default: combination of {{ @@yPercent }} ({{ @@y }} Â± {{ @@yError }})

  enableLink: false,
  linkOpenNewTab: true,
  linkFormat: "", // template like a textFormat

  missingValuesAsZero: true,
};

export default function getOptions(options: any) {
  const result = merge(
    {},
    DEFAULT_OPTIONS,
    {
      showDataLabels: options.globalSeriesType === "pie",
      dateTimeFormat: visualizationsSettings.dateTimeFormat,
    },
    options
  );

  // Backward compatibility
  if (["normal", "percent"].indexOf(result.series.stacking) >= 0) {
    result.series.percentValues = result.series.stacking === "percent";
    result.series.stacking = "stack";
  }

  return result;
}
