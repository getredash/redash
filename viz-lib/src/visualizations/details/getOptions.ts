import _ from "lodash";
import {
  getDefaultFormatOptions,
  getColumnsOptions,
} from "@/visualizations/shared/columnUtils";

const DEFAULT_OPTIONS = {};


export default function getOptions(options: any, { columns }: any) {
  options = { ...DEFAULT_OPTIONS, ...options };
  options.columns = _.map(getColumnsOptions(columns, options.columns, { alignContent: "left" }), col => ({
    ...getDefaultFormatOptions(col),
    ...col,
  }));
  return options;
}
