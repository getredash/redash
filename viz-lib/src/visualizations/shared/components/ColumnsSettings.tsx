import { map } from "lodash";
import React from "react";
import Collapse from "antd/lib/collapse";
import Tooltip from "antd/lib/tooltip";
import Typography from "antd/lib/typography";
// @ts-expect-error ts-migrate(2724) FIXME: Module '"../../../../node_modules/react-sortable-h... Remove this comment to see the full error message
import { sortableElement } from "react-sortable-hoc";
import { SortableContainer, DragHandle } from "@/components/sortable";
import PropTypes from "prop-types";

import EyeOutlinedIcon from "@ant-design/icons/EyeOutlined";
import EyeInvisibleOutlinedIcon from "@ant-design/icons/EyeInvisibleOutlined";

import ColumnEditor from "./ColumnEditor";

const { Text } = Typography;

const SortableItem = sortableElement(Collapse.Panel);

type ColumnsSettingsProps = {
  options: any;
  onOptionsChange: any;
  variant: "table" | "details";
};

export default function ColumnsSettings({ options, onOptionsChange, variant }: ColumnsSettingsProps) {
  function handleColumnChange(newColumn: any, event: any) {
    if (event) {
      event.stopPropagation();
    }
    const columns = map(options.columns, c => (c.name === newColumn.name ? newColumn : c));
    onOptionsChange({ columns });
  }

  function handleColumnsReorder({ oldIndex, newIndex }: any) {
    const columns = [...options.columns];
    columns.splice(newIndex, 0, ...columns.splice(oldIndex, 1));
    onOptionsChange({ columns });
  }

  const helperClass = `${variant}-editor-columns-dragged-item`;
  const containerClass = `${variant}-visualization-editor-columns`;
  const testPrefix = variant === "table" ? "Table" : "Details";

  return (
    <SortableContainer
      axis="y"
      lockAxis="y"
      useDragHandle
      helperClass={helperClass}
      helperContainer={(container: any) => container.firstChild}
      onSortEnd={handleColumnsReorder}
      containerProps={{
        className: containerClass,
      }}>
      {/* @ts-expect-error ts-migrate(2322) FIXME: Type 'Element' is not assignable to type 'null | u... Remove this comment to see the full error message */}
      <Collapse bordered={false} defaultActiveKey={[]} expandIconPosition="right">
        {map(options.columns, (column, index) => (
          <SortableItem
            key={column.name}
            index={index}
            header={
              <React.Fragment>
                <DragHandle />
                <span data-test={`${testPrefix}.Column.${column.name}.Name`}>
                  {column.name}
                  {column.title !== "" && column.title !== column.name && (
                    <Text type="secondary" style={{ marginLeft: 5 }}>
                      <i>({column.title})</i>
                    </Text>
                  )}
                </span>
              </React.Fragment>
            }
            extra={
              <Tooltip title="Toggle visibility" mouseEnterDelay={0} mouseLeaveDelay={0}>
                {column.visible ? (
                  <EyeOutlinedIcon
                    data-test={`${testPrefix}.Column.${column.name}.Visibility`}
                    onClick={event => handleColumnChange({ ...column, visible: !column.visible }, event)}
                  />
                ) : (
                  <EyeInvisibleOutlinedIcon
                    data-test={`${testPrefix}.Column.${column.name}.Visibility`}
                    onClick={event => handleColumnChange({ ...column, visible: !column.visible }, event)}
                  />
                )}
              </Tooltip>
            }>
            <ColumnEditor column={column} variant={variant} onChange={(changes) => handleColumnChange(changes, undefined)} />
          </SortableItem>
        ))}
      </Collapse>
    </SortableContainer>
  );
}

ColumnsSettings.propTypes = {
  options: PropTypes.object.isRequired,
  onOptionsChange: PropTypes.func.isRequired,
  variant: PropTypes.oneOf(["table", "details"]).isRequired,
};
