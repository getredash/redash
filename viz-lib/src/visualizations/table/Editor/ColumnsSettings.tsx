import React from "react";
import SharedColumnsSettings from "../../shared/components/ColumnsSettings";
import { EditorPropTypes } from "@/visualizations/prop-types";

export default function ColumnsSettings({ options, onOptionsChange, data }: any) {
  return (
    <SharedColumnsSettings
      options={options}
      onOptionsChange={onOptionsChange}
      variant="table"
    />
  );
}

ColumnsSettings.propTypes = EditorPropTypes;
