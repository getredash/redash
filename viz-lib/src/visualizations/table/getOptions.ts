import _ from "lodash";
import {
  getDefaultColumnsOptions,
  getDefaultFormatOptions,
  getColumnsOptions,
} from "@/visualizations/shared/columnUtils";

const DEFAULT_OPTIONS = {
  itemsPerPage: 25,
  paginationSize: "default", // not editable through Editor
};


export default function getOptions(options: any, { columns }: any) {
  options = { ...DEFAULT_OPTIONS, ...options };
  options.columns = _.map(getColumnsOptions(columns, options.columns, { allowSearch: false }), col => ({
    ...getDefaultFormatOptions(col),
    ...col,
  }));
  return options;
}
