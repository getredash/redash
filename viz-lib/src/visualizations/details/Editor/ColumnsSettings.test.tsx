import React from "react";
import enzyme from "enzyme";

import getOptions from "../getOptions";
import ColumnsSettings from "./ColumnsSettings";

function findByTestID(wrapper: any, testId: any) {
  return wrapper.find(`[data-test="${testId}"]`);
}

function mount(options: any, done: any) {
  const data = {
    columns: [
      { name: "id", type: "integer" },
      { name: "name", type: "string" },
      { name: "created_at", type: "datetime" },
    ],
    rows: [{ id: 1, name: "test", created_at: "2023-01-01T00:00:00Z" }],
  };
  options = getOptions(options, data);
  return enzyme.mount(
    <ColumnsSettings
      visualizationName="Details"
      data={data}
      options={options}
      onOptionsChange={changedOptions => {
        expect(changedOptions).toMatchSnapshot();
        done();
      }}
    />
  );
}

describe("Visualizations -> Details -> Editor -> Columns Settings", () => {
  test("Toggles column visibility", done => {
    const el = mount({}, done);

    findByTestID(el, "Details.Column.id.Visibility")
      .last()
      .simulate("click");
  });

  test("Changes column title", done => {
    const el = mount({}, done);
    findByTestID(el, "Details.Column.name.Name")
      .last()
      .simulate("click"); // expand settings

    findByTestID(el, "Details.Column.name.Title")
      .last()
      .simulate("change", { target: { value: "Full Name" } });
  });

  test("Changes column alignment", done => {
    const el = mount({}, done);
    findByTestID(el, "Details.Column.id.Name")
      .last()
      .simulate("click"); // expand settings

    findByTestID(el, "Details.Column.id.TextAlignment")
      .last()
      .find('[data-test="TextAlignmentSelect.Center"] input')
      .simulate("change", { target: { checked: true } });
  });

  test("Changes column description", done => {
    const el = mount({}, done);
    findByTestID(el, "Details.Column.name.Name")
      .last()
      .simulate("click"); // expand settings

    findByTestID(el, "Details.Column.name.Description")
      .last()
      .simulate("change", { target: { value: "User full name" } });
  });

  test("Changes column display type", done => {
    const el = mount({}, done);
    findByTestID(el, "Details.Column.created_at.Name")
      .last()
      .simulate("click"); // expand settings

    findByTestID(el, "Details.Column.created_at.DisplayAs")
      .last()
      .simulate("mouseDown");
    findByTestID(el, "Details.Column.created_at.DisplayAs.string")
      .last()
      .simulate("click");
  });

  test("Hides multiple columns", done => {
    const el = mount({}, done);

    findByTestID(el, "Details.Column.id.Visibility")
      .last()
      .simulate("click");
  });
});
