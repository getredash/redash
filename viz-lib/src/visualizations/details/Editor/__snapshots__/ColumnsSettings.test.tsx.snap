// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Visualizations -> Details -> Editor -> Columns Settings Changes column alignment 1`] = `
Object {
  "columns": Array [
    Object {
      "alignContent": "center",
      "allowHTML": false,
      "booleanValues": Array [
        "false",
        "true",
      ],
      "dateTimeFormat": undefined,
      "description": "",
      "displayAs": "number",
      "highlightLinks": false,
      "imageHeight": "",
      "imageTitleTemplate": "{{ @ }}",
      "imageUrlTemplate": "{{ @ }}",
      "imageWidth": "",
      "linkOpenInNewTab": true,
      "linkTextTemplate": "{{ @ }}",
      "linkTitleTemplate": "{{ @ }}",
      "linkUrlTemplate": "{{ @ }}",
      "name": "id",
      "nullValue": "null",
      "numberFormat": "0,0",
      "order": 100000,
      "title": "id",
      "type": "integer",
      "visible": true,
    },
    Object {
      "alignContent": "left",
      "allowHTML": false,
      "booleanValues": Array [
        "false",
        "true",
      ],
      "dateTimeFormat": undefined,
      "description": "",
      "displayAs": "string",
      "highlightLinks": false,
      "imageHeight": "",
      "imageTitleTemplate": "{{ @ }}",
      "imageUrlTemplate": "{{ @ }}",
      "imageWidth": "",
      "linkOpenInNewTab": true,
      "linkTextTemplate": "{{ @ }}",
      "linkTitleTemplate": "{{ @ }}",
      "linkUrlTemplate": "{{ @ }}",
      "name": "name",
      "nullValue": "null",
      "numberFormat": undefined,
      "order": 100001,
      "title": "name",
      "type": "string",
      "visible": true,
    },
    Object {
      "alignContent": "right",
      "allowHTML": false,
      "booleanValues": Array [
        "false",
        "true",
      ],
      "dateTimeFormat": "DD/MM/YYYY HH:mm",
      "description": "",
      "displayAs": "datetime",
      "highlightLinks": false,
      "imageHeight": "",
      "imageTitleTemplate": "{{ @ }}",
      "imageUrlTemplate": "{{ @ }}",
      "imageWidth": "",
      "linkOpenInNewTab": true,
      "linkTextTemplate": "{{ @ }}",
      "linkTitleTemplate": "{{ @ }}",
      "linkUrlTemplate": "{{ @ }}",
      "name": "created_at",
      "nullValue": "null",
      "numberFormat": undefined,
      "order": 100002,
      "title": "created_at",
      "type": "datetime",
      "visible": true,
    },
  ],
}
`;

exports[`Visualizations -> Details -> Editor -> Columns Settings Changes column description 1`] = `
Object {
  "columns": Array [
    Object {
      "alignContent": "right",
      "allowHTML": false,
      "booleanValues": Array [
        "false",
        "true",
      ],
      "dateTimeFormat": undefined,
      "description": "",
      "displayAs": "number",
      "highlightLinks": false,
      "imageHeight": "",
      "imageTitleTemplate": "{{ @ }}",
      "imageUrlTemplate": "{{ @ }}",
      "imageWidth": "",
      "linkOpenInNewTab": true,
      "linkTextTemplate": "{{ @ }}",
      "linkTitleTemplate": "{{ @ }}",
      "linkUrlTemplate": "{{ @ }}",
      "name": "id",
      "nullValue": "null",
      "numberFormat": "0,0",
      "order": 100000,
      "title": "id",
      "type": "integer",
      "visible": true,
    },
    Object {
      "alignContent": "left",
      "allowHTML": false,
      "booleanValues": Array [
        "false",
        "true",
      ],
      "dateTimeFormat": undefined,
      "description": "User full name",
      "displayAs": "string",
      "highlightLinks": false,
      "imageHeight": "",
      "imageTitleTemplate": "{{ @ }}",
      "imageUrlTemplate": "{{ @ }}",
      "imageWidth": "",
      "linkOpenInNewTab": true,
      "linkTextTemplate": "{{ @ }}",
      "linkTitleTemplate": "{{ @ }}",
      "linkUrlTemplate": "{{ @ }}",
      "name": "name",
      "nullValue": "null",
      "numberFormat": undefined,
      "order": 100001,
      "title": "name",
      "type": "string",
      "visible": true,
    },
    Object {
      "alignContent": "right",
      "allowHTML": false,
      "booleanValues": Array [
        "false",
        "true",
      ],
      "dateTimeFormat": "DD/MM/YYYY HH:mm",
      "description": "",
      "displayAs": "datetime",
      "highlightLinks": false,
      "imageHeight": "",
      "imageTitleTemplate": "{{ @ }}",
      "imageUrlTemplate": "{{ @ }}",
      "imageWidth": "",
      "linkOpenInNewTab": true,
      "linkTextTemplate": "{{ @ }}",
      "linkTitleTemplate": "{{ @ }}",
      "linkUrlTemplate": "{{ @ }}",
      "name": "created_at",
      "nullValue": "null",
      "numberFormat": undefined,
      "order": 100002,
      "title": "created_at",
      "type": "datetime",
      "visible": true,
    },
  ],
}
`;

exports[`Visualizations -> Details -> Editor -> Columns Settings Changes column display type 1`] = `
Object {
  "columns": Array [
    Object {
      "alignContent": "right",
      "allowHTML": false,
      "booleanValues": Array [
        "false",
        "true",
      ],
      "dateTimeFormat": undefined,
      "description": "",
      "displayAs": "number",
      "highlightLinks": false,
      "imageHeight": "",
      "imageTitleTemplate": "{{ @ }}",
      "imageUrlTemplate": "{{ @ }}",
      "imageWidth": "",
      "linkOpenInNewTab": true,
      "linkTextTemplate": "{{ @ }}",
      "linkTitleTemplate": "{{ @ }}",
      "linkUrlTemplate": "{{ @ }}",
      "name": "id",
      "nullValue": "null",
      "numberFormat": "0,0",
      "order": 100000,
      "title": "id",
      "type": "integer",
      "visible": true,
    },
    Object {
      "alignContent": "left",
      "allowHTML": false,
      "booleanValues": Array [
        "false",
        "true",
      ],
      "dateTimeFormat": undefined,
      "description": "",
      "displayAs": "string",
      "highlightLinks": false,
      "imageHeight": "",
      "imageTitleTemplate": "{{ @ }}",
      "imageUrlTemplate": "{{ @ }}",
      "imageWidth": "",
      "linkOpenInNewTab": true,
      "linkTextTemplate": "{{ @ }}",
      "linkTitleTemplate": "{{ @ }}",
      "linkUrlTemplate": "{{ @ }}",
      "name": "name",
      "nullValue": "null",
      "numberFormat": undefined,
      "order": 100001,
      "title": "name",
      "type": "string",
      "visible": true,
    },
    Object {
      "alignContent": "right",
      "allowHTML": false,
      "booleanValues": Array [
        "false",
        "true",
      ],
      "dateTimeFormat": "DD/MM/YYYY HH:mm",
      "description": "",
      "displayAs": "string",
      "highlightLinks": false,
      "imageHeight": "",
      "imageTitleTemplate": "{{ @ }}",
      "imageUrlTemplate": "{{ @ }}",
      "imageWidth": "",
      "linkOpenInNewTab": true,
      "linkTextTemplate": "{{ @ }}",
      "linkTitleTemplate": "{{ @ }}",
      "linkUrlTemplate": "{{ @ }}",
      "name": "created_at",
      "nullValue": "null",
      "numberFormat": undefined,
      "order": 100002,
      "title": "created_at",
      "type": "datetime",
      "visible": true,
    },
  ],
}
`;

exports[`Visualizations -> Details -> Editor -> Columns Settings Changes column title 1`] = `
Object {
  "columns": Array [
    Object {
      "alignContent": "right",
      "allowHTML": false,
      "booleanValues": Array [
        "false",
        "true",
      ],
      "dateTimeFormat": undefined,
      "description": "",
      "displayAs": "number",
      "highlightLinks": false,
      "imageHeight": "",
      "imageTitleTemplate": "{{ @ }}",
      "imageUrlTemplate": "{{ @ }}",
      "imageWidth": "",
      "linkOpenInNewTab": true,
      "linkTextTemplate": "{{ @ }}",
      "linkTitleTemplate": "{{ @ }}",
      "linkUrlTemplate": "{{ @ }}",
      "name": "id",
      "nullValue": "null",
      "numberFormat": "0,0",
      "order": 100000,
      "title": "id",
      "type": "integer",
      "visible": true,
    },
    Object {
      "alignContent": "left",
      "allowHTML": false,
      "booleanValues": Array [
        "false",
        "true",
      ],
      "dateTimeFormat": undefined,
      "description": "",
      "displayAs": "string",
      "highlightLinks": false,
      "imageHeight": "",
      "imageTitleTemplate": "{{ @ }}",
      "imageUrlTemplate": "{{ @ }}",
      "imageWidth": "",
      "linkOpenInNewTab": true,
      "linkTextTemplate": "{{ @ }}",
      "linkTitleTemplate": "{{ @ }}",
      "linkUrlTemplate": "{{ @ }}",
      "name": "name",
      "nullValue": "null",
      "numberFormat": undefined,
      "order": 100001,
      "title": "Full Name",
      "type": "string",
      "visible": true,
    },
    Object {
      "alignContent": "right",
      "allowHTML": false,
      "booleanValues": Array [
        "false",
        "true",
      ],
      "dateTimeFormat": "DD/MM/YYYY HH:mm",
      "description": "",
      "displayAs": "datetime",
      "highlightLinks": false,
      "imageHeight": "",
      "imageTitleTemplate": "{{ @ }}",
      "imageUrlTemplate": "{{ @ }}",
      "imageWidth": "",
      "linkOpenInNewTab": true,
      "linkTextTemplate": "{{ @ }}",
      "linkTitleTemplate": "{{ @ }}",
      "linkUrlTemplate": "{{ @ }}",
      "name": "created_at",
      "nullValue": "null",
      "numberFormat": undefined,
      "order": 100002,
      "title": "created_at",
      "type": "datetime",
      "visible": true,
    },
  ],
}
`;

exports[`Visualizations -> Details -> Editor -> Columns Settings Hides multiple columns 1`] = `
Object {
  "columns": Array [
    Object {
      "alignContent": "right",
      "allowHTML": false,
      "booleanValues": Array [
        "false",
        "true",
      ],
      "dateTimeFormat": undefined,
      "description": "",
      "displayAs": "number",
      "highlightLinks": false,
      "imageHeight": "",
      "imageTitleTemplate": "{{ @ }}",
      "imageUrlTemplate": "{{ @ }}",
      "imageWidth": "",
      "linkOpenInNewTab": true,
      "linkTextTemplate": "{{ @ }}",
      "linkTitleTemplate": "{{ @ }}",
      "linkUrlTemplate": "{{ @ }}",
      "name": "id",
      "nullValue": "null",
      "numberFormat": "0,0",
      "order": 100000,
      "title": "id",
      "type": "integer",
      "visible": false,
    },
    Object {
      "alignContent": "left",
      "allowHTML": false,
      "booleanValues": Array [
        "false",
        "true",
      ],
      "dateTimeFormat": undefined,
      "description": "",
      "displayAs": "string",
      "highlightLinks": false,
      "imageHeight": "",
      "imageTitleTemplate": "{{ @ }}",
      "imageUrlTemplate": "{{ @ }}",
      "imageWidth": "",
      "linkOpenInNewTab": true,
      "linkTextTemplate": "{{ @ }}",
      "linkTitleTemplate": "{{ @ }}",
      "linkUrlTemplate": "{{ @ }}",
      "name": "name",
      "nullValue": "null",
      "numberFormat": undefined,
      "order": 100001,
      "title": "name",
      "type": "string",
      "visible": true,
    },
    Object {
      "alignContent": "right",
      "allowHTML": false,
      "booleanValues": Array [
        "false",
        "true",
      ],
      "dateTimeFormat": "DD/MM/YYYY HH:mm",
      "description": "",
      "displayAs": "datetime",
      "highlightLinks": false,
      "imageHeight": "",
      "imageTitleTemplate": "{{ @ }}",
      "imageUrlTemplate": "{{ @ }}",
      "imageWidth": "",
      "linkOpenInNewTab": true,
      "linkTextTemplate": "{{ @ }}",
      "linkTitleTemplate": "{{ @ }}",
      "linkUrlTemplate": "{{ @ }}",
      "name": "created_at",
      "nullValue": "null",
      "numberFormat": undefined,
      "order": 100002,
      "title": "created_at",
      "type": "datetime",
      "visible": true,
    },
  ],
}
`;

exports[`Visualizations -> Details -> Editor -> Columns Settings Toggles column visibility 1`] = `
Object {
  "columns": Array [
    Object {
      "alignContent": "right",
      "allowHTML": false,
      "booleanValues": Array [
        "false",
        "true",
      ],
      "dateTimeFormat": undefined,
      "description": "",
      "displayAs": "number",
      "highlightLinks": false,
      "imageHeight": "",
      "imageTitleTemplate": "{{ @ }}",
      "imageUrlTemplate": "{{ @ }}",
      "imageWidth": "",
      "linkOpenInNewTab": true,
      "linkTextTemplate": "{{ @ }}",
      "linkTitleTemplate": "{{ @ }}",
      "linkUrlTemplate": "{{ @ }}",
      "name": "id",
      "nullValue": "null",
      "numberFormat": "0,0",
      "order": 100000,
      "title": "id",
      "type": "integer",
      "visible": false,
    },
    Object {
      "alignContent": "left",
      "allowHTML": false,
      "booleanValues": Array [
        "false",
        "true",
      ],
      "dateTimeFormat": undefined,
      "description": "",
      "displayAs": "string",
      "highlightLinks": false,
      "imageHeight": "",
      "imageTitleTemplate": "{{ @ }}",
      "imageUrlTemplate": "{{ @ }}",
      "imageWidth": "",
      "linkOpenInNewTab": true,
      "linkTextTemplate": "{{ @ }}",
      "linkTitleTemplate": "{{ @ }}",
      "linkUrlTemplate": "{{ @ }}",
      "name": "name",
      "nullValue": "null",
      "numberFormat": undefined,
      "order": 100001,
      "title": "name",
      "type": "string",
      "visible": true,
    },
    Object {
      "alignContent": "right",
      "allowHTML": false,
      "booleanValues": Array [
        "false",
        "true",
      ],
      "dateTimeFormat": "DD/MM/YYYY HH:mm",
      "description": "",
      "displayAs": "datetime",
      "highlightLinks": false,
      "imageHeight": "",
      "imageTitleTemplate": "{{ @ }}",
      "imageUrlTemplate": "{{ @ }}",
      "imageWidth": "",
      "linkOpenInNewTab": true,
      "linkTextTemplate": "{{ @ }}",
      "linkTitleTemplate": "{{ @ }}",
      "linkUrlTemplate": "{{ @ }}",
      "name": "created_at",
      "nullValue": "null",
      "numberFormat": undefined,
      "order": 100002,
      "title": "created_at",
      "type": "datetime",
      "visible": true,
    },
  ],
}
`;
